RM4 support [WIP]

ynuyasha opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->
RM4 support [WIP]
<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific addon: Mention the addon shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->
Add support for rm4 devices described in https://github.com/mjg59/python-broadlink/blob/0.13.0/broadlink/__init__.py#L50
<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the addon README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It's a good practice to add an URL to your built JAR in this Pull Request description,
so it's easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it's reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

ynuyasha commented:
can you try to build binding from my repo and try if works for your device?
I resolved some bugs on authentication

tknaller commented:
just tried, it happens as I had the problem before:
```
08:37:02.135 [ERROR] [adlink.handler.BroadlinkRemoteHandler] - rm3:c8b2c2c2[?]: java.net.ProtocolException: Response from device is not valid. (0x22=0xF9,0x23=0xFF,0x24=0xFF)
08:37:02.138 [ERROR] [adlink.handler.BroadlinkRemoteHandler] - rm3:c8b2c2c2[?]: ANTOBUG: linea 268 Basethinghandler
08:37:02.139 [ERROR] [adlink.handler.BroadlinkRemoteHandler] - rm3:c8b2c2c2[?]: Attempting to authenticate prior to getting device status FAILED. Will mark as offline
08:37:02.140 [ERROR] [adlink.handler.BroadlinkRemoteHandler] - rm3:c8b2c2c2[?]: updateItemStatus: Online -> Offline
```
When replacing the `clearProperty `again with `setProperty(bla,null)` it works...I can't tell you anythign else :)

ynuyasha commented:
I reverted to `setProperty(bla,null)` and added `setProperty(String propName, @Nullable String propValue) {`
But build failed with this error
```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project org.openhab.binding.broadlink: Compilation failure
[ERROR] openhab2-addons\openhab2-addons\bundles\org.openhab.binding.broadlink\src\main\java\org\openhab\binding\broadlink\handler\BroadlinkBaseThingHandler.java:[298,11772] Null type mismatch (type annotations): required '@NonNull String' but this expression has type '@Nullable String'
```

this is my env:
```
mvn -v
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\blabla\apache-maven-3.6.3\bin\..
Java version: 1.8.0_242, vendor: AdoptOpenJDK, runtime: C:\Program Files\AdoptOpenJDK\jdk-8.0.242.08-hotspot\jre
Default locale: it_IT, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

tknaller commented:
this is my env:
```
$ mvn -v
Apache Maven 3.6.0
Maven home: /usr/share/maven
Java version: 11.0.6, vendor: Ubuntu, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "4.4.59+", arch: "amd64", family: "unix"
```

apparantly my ubuntu has multiple JDKs and is using ver 11:
```
$ dpkg --list | grep jdk
ii  openjdk-11-jre-headless:amd64         11.0.6+10-1ubuntu1~18.04.1          amd64        OpenJDK Java runtime, using Hotspot JIT (headless)
ii  openjdk-8-jdk:amd64                   8u242-b08-0ubuntu3~18.04            amd64        OpenJDK Development Kit (JDK)
ii  openjdk-8-jdk-headless:amd64          8u242-b08-0ubuntu3~18.04            amd64        OpenJDK Development Kit (JDK) (headless)
ii  openjdk-8-jre:amd64                   8u242-b08-0ubuntu3~18.04            amd64        OpenJDK Java runtime, using Hotspot JIT
ii  openjdk-8-jre-headless:amd64          8u242-b08-0ubuntu3~18.04            amd64        OpenJDK Java runtime, using Hotspot JIT (headless)
``` 

that could be the difference :)

ynuyasha commented:
i set maven to use java 11 but the result is the same on both java 8 or 11

```
mvn -v
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: apache-maven-3.6.3\bin\..
Java version: 11.0.6, vendor: AdoptOpenJDK, runtime: C:\Program Files\AdoptOpenJDK\jdk-11.0.6.10-hotspot
Default locale: it_IT, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project org.openhab.binding.broadlink: Compilation failure
[ERROR] openhab2-addons\openhab2-addons\bundles\org.openhab.binding.broadlink\src\main\java\org\openhab\binding\broadlink\handler\BroadlinkBaseThingHandler.java:[298,11772] Null type mismatch (type annotations): required '@NonNull String' but this expression has type '@Nullable String'
```

tknaller commented:
haha awesome that goes beyond my java knowledge 😄  sorry

ynuyasha commented on bundles/org.openhab.binding.broadlink/src/main/java/org/openhab/binding/broadlink/handler/BroadlinkBaseThingHandler.java:
```diff
@@ -166,16 +170,20 @@ protected boolean authenticate() {
-        if (properties.get("key") == null || properties.get("id") == null) {
+        if (StringUtils.isEmpty(properties.get("key"))) {
+            
             key = Hex.convertHexToBytes(thingConfig.getAuthorizationKey());
```
i found a strange behavior often key is empty

ynuyasha replied:
if i restart openhab key was get correctly from thing

ynuyasha commented:
i have rollback to setproperty and add some trace logs. Can you try and tell me something about key variable too?

tknaller commented:
yeah that works 👍
